compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPackSharedMemoryAlloc.cpp

```diff
@@ -0,0 +1,225 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+
+#include "iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/NVGPU/IR/NVGPUDialect.h"
+#include "mlir/IR/Dominance.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-pack-shared-memory-alloc"
+
+namespace mlir {
+namespace iree_compiler {
+
+using AliasGroup = SmallVector<Operation *>;
+
+/// Analyze the liverange of allocs and set them in individual groups if
+/// possible and mark allocs with an attributes.
+/// The algorithm is a simplistic memory allocation solution. It sorts
+/// allocations into alias groups. Everytime two alloc's liverange interfers
+/// they are merge into the same group. If a new alloc is part of multiple alias
+/// groups all those are merged into one. At the end we are left with groups of
+/// allocations that are disjoint and can use the same memory.
+// TODO: Move this to a common place if needed by other backends.
+static void analyzeSharedMemoryAlloc(func::FuncOp funcOp,
```

Comment:
Nit (but not really) : I think the analysis here and the rewrite can be made a general utility function (even if it isnt moved yet into a different file, but ideally this could be moved to `Codegen/Utils` or `Codegen\Transforms`, i'd prefer `Transforms`.), and it is better to do it now while this landing to make sure we have clear separation of concerns (harder to untangle later). Ill leave comments related to this in other places, but there is nothing specific to shared memory in this analysis (especially since the allocations list is provided as argument).